doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,413 @@
+  * What type gets reported in the system catalog for DLQ tables? Is it really a
+    table? Or is it a source or a subsource?
+  * Is it useful to report the `key` and the `value` in the DLQ relation? Most
+    of the time they will be unreadable bytes.
+  * Can the `REDIRECT ERRORS` option be added to an existing source (i.e., via
+    `ALTER SOURCE`)? If so, what is the behavior? Does it pick up for all new
+    messages?
+  * Does the `mz_timestamp` column make sense? Is there something better
+    to call this column? Do the semantics make sense? (Including an
+    `ingested_at` wall clock timestamp would be nice, but that wouldn't
+    be a deterministic function of reclocking.)
+
+#### Upsert errors
+
+While the DLQ is useful for alerting users to data quality issues, it doesn't
+help users answer the question "does my upsert source currently have any keys
+whose most recent value failed to decode?"
+
+To solve this problem, we propose the addition of an `INLINE ERRORS` option
+to `ENVELOPE UPSERT`:
+
+```sql
+CREATE SOURCE src TO KAFKA CONNECTION kconn ...
+ENVELOPE UPSERT (
+    VALUE DECODING ERRORS = ({INLINE | PROPAGATE}, ...)
+)
+```
+
+The default behavior, which matches today's behavior, is `PROPAGATE`. Value
+decoding errors are propagated to the DLQ table or the source's *err stream*,
+whichever is active.
+
+When the `INLINE` behavior is specified, the source's relation gains a nullable
+column named `error` with a type of `record(description: text, code: text)`. If
+the most recent value for a key has been successfully decoded, this column will
+be `NULL`. If the most recent value for a key was not succesfully decoded, this
+column will contain details about the error. Additionally, the value is forced
+into a single nullable column named `value` with a type reflecting the
+configured value format—flattening of record-type values into top-level columns
+does not occur.
+
+When the `PROPAGATE` behavior is not specified, value decoding errors are
+essentially *not* treated as errors. They are neither forwarded to the DLQ table
+nor the source *errs stream*.
+
+When both `INLINE` and `PROPAGATE` are specified, the errors are both reported
+inline in the source and propagated to the DLQ table or the source's *errs
+stream*, whichever is active.
+
+Even when using `VALUE DECODING ERRORS = INLINE`, users need to monitor the DLQ
+table or the source's *errs stream* for errors, as errors while decoding the key
+still get sent to the DLQ. (There is no good way to represent key decoding
+errors inline without breaking upsert semantics.)
+
+The `INLINE` behavior allows users to discover upsert sources that presently
+have bad data by querying each source and checking `count(error) > 0`. They can
+discover the affected keys by running `SELECT key ... WHERE error IS NOT NULL`.
+
+**Open questions:**
+
+  * Does the separation of the `REDIRECT ERRORS` and the `VALUE DECODING ERRORS`
```

Comment:
I'm also slightly confused on why we have both the DLQ and the inline errors feature to solve this issue. If we're already adding two new columns to support the INLINE feature for upsert sources, why not just add those same two columns for append-only sources too? And then expose one option to treat decode errors as either real errors or inline errors, with no DLQ at all?

The downside is that append-only source values would become nested in this case, but this seems like it could be an acceptable tradeoff for folks who need inline error support?

Reply:
> If we're already adding two new columns to support the INLINE feature for upsert sources, why not just add those same two columns for append-only sources too? ... The downside is that append-only source values would become nested in this case, but this seems like it could be an acceptable tradeoff for folks who need inline error support?

Agreed that this works nicely for `ENVELOPE NONE` sources, and agree that the nesting of the `values` column seems acceptable.

The place where I got stuck is: where do you put errors for `ENVELOPE UPSERT` that occur when decoding the key? Without a DLQ there isn't anywhere good to put them.

Reply:
> For an upsert source, it can't tell you whether all keys currently have valid values or not.

Similar to above, I think we could punt on this. I think the V0 deliverable is "sources don't break when bad data shows up," and the V1+ deliverables can improve developer workflows when something _is_ broken.

Can we cut this from scope? Or would we be painting ourselves into a corner?

Reply:
referencing @benesch 's follow up comment here: https://github.com/MaterializeInc/materialize/pull/27540#issuecomment-2174340357

It sounds like we're all thinking that we could get away with just 1 option for the MVP and satisfy this customer's issue, but depends on which. @JLDLaughlin suggested cutting the inline-error feature and just delivering the DLQ, and @benesch is suggesting cutting the DLQ and just supporting the inline-errors which solve the primary issue of value decode errors in upsert sources.

I'm on the fence -- the DLQ handles more cases overall but it means an upsert source may present incorrect data (a stale value for a key), whereas the inline-error option provides correctness but doesnt account for key-decode errors and append-only sources...

Reply:
Discussed this live with Jessica, since I've been thinking we might want to go in the _other_ direction and _only_ do `VALUE DECODING ERRORS = INLINE`.